### PR TITLE
Get the new CVR import functions in libs/backend up to 100% code coverage

### PR DESCRIPTION
**Pull request by arsalansufi**
_Easiest reviewed by commit_

## Overview

Issue link: https://github.com/votingworks/vxsuite/issues/4058

The counterpart to https://github.com/votingworks/vxsuite/pull/4090, this PR gets the new CVR import functions in `libs/backend` up to 💯% code coverage. It also adds one e2e test of export and subsequent import of that export.

### CODE REVIEW

File: libs/backend/src/cast_vote_records/import.ts [hunk 1]
```diff
@@ -167,10 +167,8 @@ async function* castVoteRecordGenerator(
...
         !castVoteRecord.BallotImage[0].Hash?.Value ||
         !castVoteRecord.BallotImage[1].Hash?.Value ||
-        // These next two conditions are hard to trigger without first triggering the invalid
-        // write-in case above.
-        /* istanbul ignore next */ !castVoteRecord.BallotImage[0].Location ||
-        /* istanbul ignore next */ !castVoteRecord.BallotImage[1].Location ||
+        !castVoteRecord.BallotImage[0].Location ||
+        !castVoteRecord.BallotImage[1].Location ||
         !castVoteRecord.BallotImage[0].Location.startsWith('file:') ||
         !castVoteRecord.BallotImage[1].Location.startsWith('file:')
```
Comment (arsalansufi) [inline libs/backend/src/cast_vote_records/import.ts:173]: Some interesting behavior here that I haven't fully grok'd, but I actually only needed the ignores when I was using optional chaining here, before this PR, e.g.

coverage

When I get rid of the optional chaining, I hit 100% coverage, no ignores or updates to tests needed.

File: libs/backend/src/cast_vote_records/export.test.ts [hunk 2]
```diff
@@ -1028,3 +1031,50 @@ test('areOrWereCastVoteRecordsBeingExportedToUsbDrive returns false when no USB
...
+  // Import
+  process.env['VX_MACHINE_TYPE'] = 'admin';
+  const exportDirectoryPaths = await getCastVoteRecordExportDirectoryPaths(
+    mockUsbDrive.usbDrive
+  );
+  expect(exportDirectoryPaths).toHaveLength(1);
+  const exportDirectoryPath = assertDefined(exportDirectoryPaths[0]);
+  const readResult = await readCastVoteRecordExport(exportDirectoryPath);
+  expect(readResult.isOk()).toEqual(true);
+  assert(readResult.isOk());
```
Comment (eventualbuddha) [inline libs/backend/src/cast_vote_records/export.test.ts:1071]: Just keep the `assert`?
```suggestion
  assert(readResult.isOk());
```

Or alternatively change the next line to this and remove the assert/expect:

```ts
  const { castVoteRecordIterator } = readResult.unsafeUnwrap();
```

File: libs/backend/src/cast_vote_records/import.test.ts [hunk 3]
```diff
@@ -1,7 +1,325 @@
...
+const { castVoteRecordExport } = electionGridLayoutNewHampshireAmherstFixtures;
+
+test('successful import', async () => {
+  const exportDirectoryPath = castVoteRecordExport.asDirectoryPath();
+
+  const readResult = await readCastVoteRecordExport(exportDirectoryPath);
+  expect(readResult.isOk()).toEqual(true);
+  assert(readResult.isOk());
+  const { castVoteRecordExportMetadata, castVoteRecordIterator } =
+    readResult.ok();
```
Comment (eventualbuddha) [inline libs/backend/src/cast_vote_records/import.test.ts:44]: Ditto here.

File: libs/backend/src/cast_vote_records/import.test.ts [hunk 4]
```diff
@@ -1,7 +1,325 @@
...
+    castVoteRecordReportMetadata: expect.any(Object),
+    castVoteRecordRootHash: expect.stringMatching(/^[a-f0-9]{64}$/),
+  });
+
+  let encounteredReferencedImageFiles = false;
+  let encounteredReferencedLayoutFiles = false;
+  for await (const castVoteRecordResult of castVoteRecordIterator) {
+    expect(castVoteRecordResult.isOk()).toEqual(true);
+    assert(castVoteRecordResult.isOk());
+    const { castVoteRecord, referencedFiles } = castVoteRecordResult.ok();
```
Comment (eventualbuddha) [inline libs/backend/src/cast_vote_records/import.test.ts:56]: Etc. I'm going to stop flagging them.

File: libs/backend/src/cast_vote_records/import.test.ts [hunk 5]
```diff
@@ -1,7 +1,325 @@
...
+    path.join(exportDirectoryPath, CastVoteRecordExportFileName.METADATA),
+    '}'
+  );
+
+  expect(await readCastVoteRecordExport(exportDirectoryPath)).toEqual(
+    err({ type: 'metadata-file-parse-error' })
+  );
+});
+
+test.each<{ modifier: (castVoteRecordReportPath: string) => void }>([
```
Comment (eventualbuddha) [inline libs/backend/src/cast_vote_records/import.test.ts:125]: Without giving these a separate description, we get a bunch of identical test names:
image

File: libs/backend/src/cast_vote_records/import.test.ts [hunk 6]
```diff
@@ -1,7 +1,325 @@
...
+
+  const { castVoteRecordIterator } = (
+    await readCastVoteRecordExport(exportDirectoryPath)
+  ).unsafeUnwrap();
+  const castVoteRecordResults = await castVoteRecordIterator.toArray();
+  expect(castVoteRecordResults).toHaveLength(1);
+  const castVoteRecordResult = assertDefined(castVoteRecordResults[0]);
+  expect(castVoteRecordResult).toEqual(
+    err({ type: 'invalid-cast-vote-record', subType: 'parse-error' })
+  );
```
Comment (eventualbuddha) [inline libs/backend/src/cast_vote_records/import.test.ts:167]: Simplify to this?
```suggestion
  expect(castVoteRecordResults).toEqual([
    err({ type: 'invalid-cast-vote-record', subType: 'parse-error' })
  ]);
```

File: libs/backend/src/cast_vote_records/export.test.ts (hunk 2 above)
Comment (arsalansufi) [reply libs/backend/src/cast_vote_records/export.test.ts:1071]: Ah yeah, I'd been using a combo of `expect` and `assert` because I find the output of `expect` when a check fails to be more user friendly than `assert`. But `expect` doesn't hint the type checker so an `assert` still remains necessary.

On second thought, though, the combo does feel redundant. Switched to `unsafeUnwrap` where I could: https://github.com/votingworks/vxsuite/pull/4091/commits/e73f2bf922941f82635f28ca6b0faa78640653b8

File: libs/backend/src/cast_vote_records/import.test.ts (hunk 5 above)
Comment (arsalansufi) [reply libs/backend/src/cast_vote_records/import.test.ts]: Yeah I sometimes get lazy about this 😅

Updated such that all test descriptions are unique: https://github.com/votingworks/vxsuite/pull/4091/commits/af139cf547e8d8c23c224d779fe5f141406c0840

File: libs/backend/src/cast_vote_records/import.test.ts (hunk 6 above)
Comment (arsalansufi) [reply libs/backend/src/cast_vote_records/import.test.ts]: :100: https://github.com/votingworks/vxsuite/pull/4091/commits/d4ed6dac3986f6285e0469fca2cede3ccbc1a215

File: libs/backend/src/cast_vote_records/import.test.ts [hunk 7]
```diff
@@ -1,7 +1,325 @@
...
+    const { castVoteRecordIterator } = (
+      await readCastVoteRecordExport(exportDirectoryPath)
+    ).unsafeUnwrap();
+    const castVoteRecordResults = await castVoteRecordIterator.toArray();
+    const castVoteRecordErrorResult = castVoteRecordResults.find(
+      (castVoteRecordResult) => castVoteRecordResult.isErr()
+    );
+    expect(castVoteRecordErrorResult).toEqual(
+      err({ type: 'invalid-cast-vote-record', subType: expectedErrorSubType })
+    );
```
[the remark this answers is not in the document]
Comment (arsalansufi) [reply libs/backend/src/cast_vote_records/import.test.ts]: Nice, TIL about `.toContainEqual`: https://github.com/votingworks/vxsuite/pull/4091/commits/d4ed6dac3986f6285e0469fca2cede3ccbc1a215

File: libs/types/src/cast_vote_records.ts [hunk 8]
```diff
@@ -74,7 +74,7 @@ export type CastVoteRecordReportWithoutMetadata = Pick<
 
 export const CastVoteRecordReportWithoutMetadataSchema: z.ZodSchema<CastVoteRecordReportWithoutMetadata> =
   z.object({
-    CVR: z.array(CVRSchema),
+    CVR: z.array(CVRSchema).optional(),
```
[the remark this answers is not in the document]
Comment (arsalansufi) [reply libs/types/src/cast_vote_records.ts:77]: Yeah a tricky edge case for sure!

File: libs/backend/src/cast_vote_records/import.test.ts [hunk 9]
```diff
@@ -1,7 +1,325 @@
...
+      numCastVoteRecordsToKeep: 10,
+    },
+    expectedErrorSubType: 'invalid-ballot-image-field',
+  },
+  {
+    description: 'invalid ballot image field',
+    modifications: {
+      castVoteRecordModifier: (castVoteRecord) =>
+        castVoteRecord.BallotImage
+          ? set(castVoteRecord, 'BallotImage[1].vxLayoutFileHash', '')
```
[the remark this answers is not in the document]
Comment (arsalansufi) [reply libs/backend/src/cast_vote_records/import.test.ts:296]: 😅 Yeah I went back and forth on this but ultimately decided that readability was a higher priority than type safety in this context
